app/bin/tools/check_domain_access.dart

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+/// The script checks if the domains accessed by the pub client are available,
+/// and not blocked by any network infrastructure.
+Future<void> main(List<String> args) async {
+  final urls = <String>[
+    // package info
+    'https://pub.dev/api/packages/http',
+    'https://pub.dartlang.org/api/packages/http',
+    // package archive
+    'https://pub.dev/packages/http/versions/0.13.1.tar.gz',
+    'https://pub.dartlang.org/packages/http/versions/0.13.1.tar.gz',
+    'https://storage.googleapis.com/pub-packages/packages/http-0.13.1.tar.gz',
+  ];
+
+  final iterations = args.isEmpty ? 10 : int.parse(args.first);
+
+  for (var i = 0; i < iterations; i++) {
+    print('Iteration #${i + 1}/$iterations...');
+    for (final url in urls) {
+      await _checkUrlAccess(url);
+    }
+  }
+  print('Done.');
+}
+
+Future<void> _checkUrlAccess(String url) async {
+  final client = HttpClient();
+  final rq = await client.getUrl(Uri.parse(url));
```

Comment:
What if we do a quick: [`lookup`](https://api.dart.dev/stable/2.12.2/dart-io/InternetAddress/lookup.html) on the address of the URL to see what it is resolved to.

Reply:
And maybe even do `Socket.connect` to see if we can reach it.

Reply:
We should probably make this tool a bit smarter... So that if the HTTP connection fails, then it prints all the IPs for the host and whether or not connecting failed.

We probably also should try all urls, even if the first one fails. But then we have a decent tool for telling people that they likely have a firewall blocking `dart`. I don't see any other reason why `dart` shouldn't be able to open TCP connections.

Short of there being some bug in the Dart SDK that prevents TCP connections from working.